03/implement-arr-to-bst.checkpoint.md

```diff
@@ -0,0 +1,281 @@
+
+    def test_will_return_balanced_bst_for_even_lengthed_list(self):
+        # Arrange
+        arr = [1, 3, 9, 27, 81, 243]
+
+        # Act
+        result = arr_to_bst(arr)
+
+        # Assert
+        self.assertTrue(self.is_bst(result))
+        self.assertTrue(self.is_balanced_tree(result))
+
+    def test_will_return_balanced_bst_for_long_list(self):
+        # Arrange
+        arr = []
+        num = 0
+        while num < 100:
+            arr.append(num)
+            num += 1
+
+        # Act
+        result = arr_to_bst(arr)
+
+        # Assert
+        self.assertTrue(self.is_bst(result))
+        self.assertTrue(self.is_balanced_tree(result))
+
+    def test_will_return_none_for_empty_list(self):
+        # Arrange
+        arr = []
+
+        # Act
+        result = arr_to_bst(arr)
+
+        # Assert
+        self.assertEqual(result, None)
+        
+
+    # Below are functions used to test if the given tree is a balanced Binary Search Tree.
+
+    # Returns True if the BST provided is a valid BST.
+    def is_bst(self, root):
+        if root is None:
+            return True
+
+        left = root.left
+        if left is not None and root.val <= left.val:
+            return False
+
+        right = root.right
+        if right is not None and root.val >= right.val:
+            return False
+
+        return self.is_bst(left) and self.is_bst(right)
+
+    # Returns the height of a tree
+    def height(self, root):
+        if root is None:
+            return 0
+        
+        left_height = self.height(root.left)
+        right_height = self.height(root.right)
+
+        if left_height > right_height:
+            return left_height + 1
+        else:
+            return right_height + 1
+
+
+    # Returns True if a tree is balanced
+    def is_balanced_tree(self, root):
+        if root is None:
+            return True
+
+        left_height = self.height(root.left)
+        right_height = self.height(root.right)
+
+        if abs(left_height - right_height) > 1:
+            return False
+
+        left_check = self.is_balanced_tree(root.left)
+        right_check = self.is_balanced_tree(root.right)
+
+        return left_check and right_check
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def arr_to_bst(arr):
+    if not arr:
+        return None
+
+    mid = (len(arr)) // 2
+
+    root = TreeNode(arr[mid])
+    root.left = arr_to_bst(arr[:mid])
+    root.right = arr_to_bst(arr[mid + 1:])
+    return root
+```
+### !end-explanation
+
+### !hint
+It is recommended to break the problem down recursively by first setting the root of the Binary Search Tree to the middle element of the array.
+### !end-hint
+
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<br>
+<details style="max-width: 700px; margin: auto;">
+<summary>Click here to see the tests that will be run against your code</summary>
+
+```py
+def test_will_return_balanced_bst_for_odd_lengthed_list(self):
```

Comment:
Same here regarding `self` in the example tests. Or since it looks like this still uses `unittest` maybe add a note about what `self` is referencing?